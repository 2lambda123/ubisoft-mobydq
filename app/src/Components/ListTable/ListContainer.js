import React from 'react';

import { Query } from 'react-apollo';
import { GraphQLError } from './../Error/GraphQLError';

import ListTable from '../ListTable/ListTable';
import LinkButton from './../../Components/FormInput/LinkButton';
import AddIcon from '@material-ui/icons/Add';

class ListContainer extends React.Component {
  renderForm(data) {
    this.props.setRowTotal(data[this.props.dataObjectName].totalCount);
    return <div>
      <div style={{ 'float': 'left', 'marginLeft': '50px' }}>
        {this.props.tableHeader}
      </div>
      <div style={{ 'float': 'right' }}>
        <LinkButton label=<AddIcon /> type="create" color="secondary" variant="fab" to={this.props.newLink}/>
      </div>
      <ListTable
        data={data[this.props.dataObjectName].nodes}
        buttons={this.props.buttons}
        {...this.props}
      />
    </div>;
  }

  render() {
    return (
      <Query
        query={this.props.repository.getListPage()}
        variables={{ 'first': this.props.rowsPerPage, 'offset': this.props.page * this.props.rowsPerPage, 'orderBy': this.props.sortColumn }}
        fetchPolicy={this.props.refetch ? 'cache-and-network' : 'cache-first'}
      >
        {({ loading, error, data }) => {
          if (loading) {
            // Check if we already have data object so we would not remove the old data before we get the new one.
            if (Object.keys(data).length) {
              return this.renderForm(data);
            }
            return <p>Loading...</p>;
          }
          if (error) {
            return <GraphQLError error={error} />;
          }
          return this.renderForm(data);
        }}
      </Query>
    );
  }
}

export default ListContainer;
