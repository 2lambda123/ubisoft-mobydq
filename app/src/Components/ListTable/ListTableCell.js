import React from 'react';
import TableCell from '@material-ui/core/TableCell';

/**
 * Table cell presentation component
 *
 * @param contents - table cell contents.
 * @param key - Table cell key
 *
 */

export const ListTableCell = ({ contents }) => <TableCell>{contents}</TableCell>;


export default ListTableCell;
