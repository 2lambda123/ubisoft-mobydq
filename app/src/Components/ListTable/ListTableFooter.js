import React from 'react';

import TableRow from '@material-ui/core/TableRow';
import TableFooter from '@material-ui/core/TableFooter';
import TablePagination from '@material-ui/core/TablePagination';
import TablePaginationActionsWrapped from './TablePaginationActions';

class ListTableFooter extends React.Component {
  constructor() {
    super();
    this.handleChangePage = this.handleChangePage.bind(this);
    this.handleChangeRowsPerPage = this.handleChangeRowsPerPage.bind(this);
  }

  render() {
    return <TableFooter>
      <TableRow>
        <TablePagination
          colSpan={10}
          count={this.props.params.rowTotal}
          rowsPerPage={this.props.params.rowsPerPage}
          page={this.props.params.page}
          onChangePage={this.handleChangePage}
          onChangeRowsPerPage={(event) => this.handleChangeRowsPerPage(event)}
          ActionsComponent={TablePaginationActionsWrapped}
        />
      </TableRow>
    </TableFooter>;
  }

  handleChangePage(page) {
    this.props.params.setPage(page);
  }

  handleChangeRowsPerPage(event) {
    this.props.params.setRowsPerPage(event.target.value);
  }
}

export default ListTableFooter;
