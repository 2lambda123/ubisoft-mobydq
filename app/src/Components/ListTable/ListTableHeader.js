import React from 'react';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import Tooltip from '@material-ui/core/Tooltip';

/**
 * Component to build the list table header.
 * @param headerNames array of column names
 */
class ListTableHeader extends React.Component {
  _handleSortClick = (fieldName, isActive, direction) => {
    const newDirection = (isActive && direction === SORT_ORDER.DESCENDING) ? 'ASC' : 'DESC';
    this.props.params.setSortColumn(fieldName + '_' + newDirection);
  };

  /**
  * GraphQL uses FIELD_NAME format for defining the sort directives as opposed to camelCase in the field names.
  * This function is responsible for the necessary field name conversion
  **/
  _buildSortableHeaderCell(header, sortHeader, sortActive, sortDirection) {
    return header === 'Actions'
      ? <React.Fragment>{header}</React.Fragment>
      : <Tooltip
          title='Sort'
          placement={'bottom-start'}
          enterDelay={300}
        >
          <TableSortLabel
            active={sortActive}
            direction={sortDirection}
            onClick={() => this._handleSortClick(sortHeader, sortActive, sortDirection)}
          >
            {header}
          </TableSortLabel>
        </Tooltip>;
  }

  _buildHeaderCell(fieldName, sortField, sortDirection) {
    let header = fieldName;
    let sortHeader = '';
   if (header.length > 0) {
      header = header.charAt(0).toUpperCase() + header.slice(1);
      sortHeader = header.match(/[A-Z][a-z]+/ug).join('_').toUpperCase();
      header = header.match(/[A-Z][a-z]+/ug).join(' ');
    }
    sortDirection = (sortField && sortHeader === sortField) ? sortDirection.toLowerCase() : SORT_ORDER.DESCENDING;
    const sortActive = sortField && sortHeader === sortField && fieldName !== 'Actions';
    return (
      <TableCell
        key={header}
      >
        {this._buildSortableHeaderCell(header, sortHeader, sortActive, sortDirection)}
      </TableCell>
    );
  }

  _buildHeader(headerFields) {
    // Add "Actions column
    const sortField = this.props.params.sortColumn;
    const allHeaderFields = headerFields.concat(['Actions']);
    //sort field is either null or FIELD_NAME_ASC/DESC
    const sortFieldName = sortField === null ? null : sortField.substring(0, sortField.lastIndexOf("_"));
    const sortDirection = sortField === null ? null : sortField.substring(sortField.lastIndexOf("_") + 1);
    return (
      <TableHead>
        <TableRow>
          {allHeaderFields.map((fieldName) => this._buildHeaderCell(fieldName, sortFieldName, sortDirection))}
        </TableRow>
      </TableHead>
    );
  }

  render() {
    return this._buildHeader(this.props.headerNames);
  }
}

const SORT_ORDER = {
    'DESCENDING': 'desc',
    'ASCENDING': 'asc'
}

export default ListTableHeader;


