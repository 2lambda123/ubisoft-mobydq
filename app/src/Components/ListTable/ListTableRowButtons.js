import React from 'react';
import { ListTableCell } from './ListTableCell';
import createEditRowFunction from './ListRowButtons/EditRowButton';
import createDeleteRowButton from './ListRowButtons/DeleteRowButton';
import createExecuteRowButton from './ListRowButtons/ExecuteRowButton';


/**
 * Container component for TableRow Action Buttons
 *
 * @param buttons a list of buttons
 * @param value an id of the row element
 * #param button.parameter base_path for a predefined function or a custom function for a custom action
 *
 * Calls a Button component to render each button
 */

class ListTableRowButtons extends React.Component {
  render() {
    return <ListTableCell contents={this.props.buttons.map((button) => this._createButton(button))} />;
  }

  _createButton(button) {
    const recordId = this.props.value;
    switch (button.function) {
      case 'edit':
        return createEditRowFunction(button, recordId);
      case 'delete':
        return createDeleteRowButton(button, recordId);
      case 'execute':
        return createExecuteRowButton(button, recordId);
      default:
        return <React.Fragment key={`none_${recordId}`} />;
    }
  }
}

export default ListTableRowButtons;
