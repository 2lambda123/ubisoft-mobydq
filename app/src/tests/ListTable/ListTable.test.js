import React from 'react'
import * as Adapter from 'enzyme-adapter-react-16';
import { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import ListTable from '../../Components/ListTable/ListTable';

describe('List Table unit test', () => {
  let wrapper;
  let buttons = [{name: 'button name', function: jest.fn}];
  let data = [{id: 1, name: 'mock name', trueField: true, falseField: false, __typename: 'Mock Type'},
              {id: 2, name: 'mock name 2', trueField: false, falseField: false, __typename: 'Mock Type'}];
  beforeEach(() => {
    wrapper = shallow(<ListTable buttons={buttons} data={data}/>)
  });

  it('renders', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});


describe('ListTable functional test', () => {
  it ('renders table correctly, including boolean fields (True/False icons', () => {
    let buttons = [{name: 'button name', function: jest.fn}];
    let data = [{id: 1, name: 'mock name', trueField: true, falseField: false, __typename: 'Mock Type'},
                {id: 2, name: 'mock name 2', trueField: false, falseField: false, __typename: 'Mock Type'}];
    let wrapper = mount(<ListTable buttons={buttons} data={data}/>);
    expect(wrapper.find('ListTableRowButtons').exists()).toBe(true);
    expect(wrapper.find('ListTableRowButtons').length).toEqual(2);
    expect(wrapper.find('ListTableRowButtons').at(0).text()).toEqual('button name');
    expect(wrapper.find('ListTableRowButtons').at(1).text()).toEqual('button name');
    expect(wrapper.find('tr').length).toEqual(3);
    expect(wrapper.find('td').length).toEqual(10);
    expect(wrapper.find('th').length).toEqual(5);
    expect(wrapper.find('th').at(0).text()).toEqual('Id');
    expect(wrapper.find('th').at(1).text()).toEqual('Name');
    expect(wrapper.find('th').at(2).text()).toEqual('True Field');
    expect(wrapper.find('th').at(3).text()).toEqual('False Field');
    expect(wrapper.find('th').at(4).text()).toEqual('Actions');
    expect(wrapper.find('th').at(0).text()).toEqual('Id');
    expect(wrapper.find('th').at(1).text()).toEqual('Name');
    expect(wrapper.find('td').at(0).text()).toEqual('1');
    expect(wrapper.find('td').at(1).text()).toEqual('mock name');
    expect(wrapper.find('td').at(5).text()).toEqual('2');
    expect(wrapper.find('td').at(6).text()).toEqual('mock name 2');
    expect(wrapper.find('DoneIcon').length).toEqual(1);
    expect(wrapper.find('ClearIcon').length).toEqual(3);
  });
  it ('renders empty table body correctly', () => {
    let buttons = [];
    let data = [];
    let wrapper = mount(<ListTable buttons={buttons} data={data}/>);
    expect(wrapper.text()).toEqual(null);
  });
});