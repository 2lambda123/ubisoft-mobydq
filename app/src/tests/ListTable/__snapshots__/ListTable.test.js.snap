// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Table unit test matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(ListTable)
    buttons={
      Array [
        Object {
          "function": "edit",
          "parameter": "/test-route",
        },
      ]
    }
    data={
      Array [
        Object {
          "__typename": "Mock Type",
          "falseField": false,
          "id": 1,
          "name": "mock name",
          "trueField": true,
        },
        Object {
          "__typename": "Mock Type",
          "falseField": false,
          "id": 2,
          "name": "mock name 2",
          "trueField": false,
        },
      ]
    }
    dataObjectName="allDataSources"
    page={1}
    repository={[Function]}
    rowTotal={2}
    rowsPerPage={10}
    setPage={[MockFunction]}
    setRowsPerPage={[MockFunction]}
    sortColumn={null}
    tableHeader="Data Sources"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttons": Array [
        Object {
          "function": "edit",
          "parameter": "/test-route",
        },
      ],
      "classes": Object {
        "appBar": "ListTable-appBar-4",
        "appBarShift": "ListTable-appBarShift-5",
        "appBarSpacer": "ListTable-appBarSpacer-11",
        "chartContainer": "ListTable-chartContainer-13",
        "content": "ListTable-content-12",
        "drawerPaper": "ListTable-drawerPaper-9",
        "drawerPaperClose": "ListTable-drawerPaperClose-10",
        "formContainer": "ListTable-formContainer-14",
        "menuButton": "ListTable-menuButton-6",
        "menuButtonHidden": "ListTable-menuButtonHidden-7",
        "root": "ListTable-root-1",
        "tableContainer": "ListTable-tableContainer-15",
        "title": "ListTable-title-8",
        "toolbar": "ListTable-toolbar-2",
        "toolbarIcon": "ListTable-toolbarIcon-3",
      },
      "data": Array [
        Object {
          "__typename": "Mock Type",
          "falseField": false,
          "id": 1,
          "name": "mock name",
          "trueField": true,
        },
        Object {
          "__typename": "Mock Type",
          "falseField": false,
          "id": 2,
          "name": "mock name 2",
          "trueField": false,
        },
      ],
      "dataObjectName": "allDataSources",
      "page": 1,
      "repository": [Function],
      "rowTotal": 2,
      "rowsPerPage": 10,
      "setPage": [MockFunction],
      "setRowsPerPage": [MockFunction],
      "sortColumn": null,
      "tableHeader": "Data Sources",
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttons": Array [
          Object {
            "function": "edit",
            "parameter": "/test-route",
          },
        ],
        "classes": Object {
          "appBar": "ListTable-appBar-4",
          "appBarShift": "ListTable-appBarShift-5",
          "appBarSpacer": "ListTable-appBarSpacer-11",
          "chartContainer": "ListTable-chartContainer-13",
          "content": "ListTable-content-12",
          "drawerPaper": "ListTable-drawerPaper-9",
          "drawerPaperClose": "ListTable-drawerPaperClose-10",
          "formContainer": "ListTable-formContainer-14",
          "menuButton": "ListTable-menuButton-6",
          "menuButtonHidden": "ListTable-menuButtonHidden-7",
          "root": "ListTable-root-1",
          "tableContainer": "ListTable-tableContainer-15",
          "title": "ListTable-title-8",
          "toolbar": "ListTable-toolbar-2",
          "toolbarIcon": "ListTable-toolbarIcon-3",
        },
        "data": Array [
          Object {
            "__typename": "Mock Type",
            "falseField": false,
            "id": 1,
            "name": "mock name",
            "trueField": true,
          },
          Object {
            "__typename": "Mock Type",
            "falseField": false,
            "id": 2,
            "name": "mock name 2",
            "trueField": false,
          },
        ],
        "dataObjectName": "allDataSources",
        "page": 1,
        "repository": [Function],
        "rowTotal": 2,
        "rowsPerPage": 10,
        "setPage": [MockFunction],
        "setRowsPerPage": [MockFunction],
        "sortColumn": null,
        "tableHeader": "Data Sources",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "childContextTypes": Object {
      "router": [Function],
    },
    "context": Object {
      "router": Object {
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        },
        "route": Object {
          "location": Object {},
          "match": Object {},
        },
      },
    },
  },
}
`;
